### ISSUE TITLE
Cargo hangs at "Resolving dependency graph..."

### ISSUE THREAD

**Issue by sorpaas** [rust-lang/cargo#11454]
### Problem

Building the following project, the build process hangs at `Resolving dependency graph...`

### Steps

1. Clone https://github.com/corepaper/morum/tree/f70f328621513dae95a4c8833878332be9211a4e
2. Run build, and it should work.
3. Uncomment line `matrix-sdk = { version = "0.6", default-features = false, features = ["e2e-encryption", "sled", "rustls-tls"] }` in `Cargo.toml` to add the dependency.
4. Cargo hangs at `Resolving dependency graph...`

### Version

Tried both stable and nightly releases:

```
cargo 1.65.0 (4bc8f24d3 2022-10-20)
cargo 1.67.0-nightly (f6e737b1e 2022-12-02)
```

**Comment by ehuss** [comment]
I suppose this is #4066 more or less. The resolver is running into conflicts, backing up and trying a different version, and then repeatedly failing. It could take hours or more for it to try all permutations.

@Eh2406 Is there anything to learn from this issue? Should it be closed in favor of #4066 (or #5284)?

**Comment by Eh2406** [comment]
Yes, my best guess is that the resolver is trying an extraordinary number of permutations without generalizing to see what the real problem is. Real-world examples are always interesting. When will I have time to follow up on this example, to see what properties make it slow? I don't know that I will make time for it, sorry.

**Comment by weihanglo** [comment]
#15837 has pointed out that passing `--offline` will immediately get a resolution. I am not sure how it works, though this might be a temporary workaround and hint to give to users.

**Comment by weihanglo** [comment]
I am going to close this as a duplicate of #4066, which has more discussions around. One of the best part of this issue is the reproduction. Thank you for providing that! I'll post a summary in #4066 then.

**Comment by weihanglo** [comment]
The output of `cargo fetch --offline` in the repro btw:

```
error: failed to select a version for `zeroize`.
    ... required by package `x25519-dalek v1.2.0`
    ... which satisfies dependency `x25519-dalek = "^1.2.0"` of package `vodozemac v0.3.0`
    ... which satisfies dependency `vodozemac = "^0.3.0"` of package `matrix-sdk-crypto v0.6.0`
    ... which satisfies dependency `matrix-sdk-crypto = "^0.6.0"` of package `matrix-sdk-sled v0.2.0`
    ... which satisfies dependency `matrix-sdk-sled = "^0.2.0"` of package `matrix-sdk v0.6.0`
    ... which satisfies dependency `matrix-sdk = "^0.6"` of package `morum v0.1.0 (/projects/morum)`
versions that meet the requirements `=1.3` are: 1.3.0

all possible versions conflict with previously selected packages.

  previously selected package `zeroize v1.8.1`
    ... which satisfies dependency `zeroize = "^1.3.0"` of package `matrix-sdk v0.6.0`
    ... which satisfies dependency `matrix-sdk = "^0.6"` of package `morum v0.1.0 (/projects/morum)`

failed to select a version for `zeroize` which could resolve this conflict
```
